### PR TITLE
ipn/ipnlocal, wgengine/netstack: use netstack for peerapi server

### PR DESCRIPTION
**Pull request by bradfitz**
We're finding a bunch of host operating systems/firewalls interact poorly
with peerapi. We either get ICMP errors from the host or users need to run
commands to allow the peerapi port:

https://github.com/tailscale/tailscale/issues/3842#issuecomment-1025133727

... even though the peerapi should be an internal implementation detail.

Rather than fight the host OS & firewalls, this change handles the
server side of peerapi entirely in netstack (except on iOS), so it
never makes its way to the host OS where it might be messed with. Two
main downsides are:

1) netstack isn't as fast, but we don't really need speed for peerapi.
   And actually, with fewer trips to/from the kernel, we might
   actually make up for some of the netstack performance loss by
   staying in userspace.

2) tcpdump / Wireshark etc packet captures will no longer see the peerapi
   traffic. Oh well. Crawshaw's been wanting to add packet capture server
   support to tailscaled, so we'll probably do that sooner now.

A future change might also then use peerapi for the client-side
(except on iOS).

Updates #3842 (probably fixes, as well as many exit node issues I bet)

Front logo [Front conversations](https://app.frontapp.com/open/top_49gpd)

### CODE REVIEW

File: ipn/ipnlocal/local.go [hunk 1]
```diff
@@ -1799,6 +1801,27 @@ func (b *LocalBackend) getPeerAPIPortForTSMPPing(ip netaddr.IP) (port uint16, ok
 	return 0, false
 }
 
+// ServePeerAPIConnection served an already-accepted connection c.
```
Comment (DentonGentry) [inline ipn/ipnlocal/local.go:1804]: ServePeerAPIConnection serves? (present tense)

File: wgengine/netstack/netstack.go [hunk 2]
```diff
@@ -82,9 +82,12 @@ type Impl struct {
...
 	logf      logger.Logf
 	dialer    *tsdial.Dialer
-	ctx       context.Context    // alive until Close
-	ctxCancel context.CancelFunc // called on Close
-	lb        *ipnlocal.LocalBackend
+	ctx       context.Context        // alive until Close
+	ctxCancel context.CancelFunc     // called on Close
+	lb        *ipnlocal.LocalBackend // or nil
+
+	peerapi4Atomic uint32 // of uint16
```
Comment (DentonGentry) [inline wgengine/netstack/netstack.go:89]: I don't understand what "of uint16" is intended to convey. This is a TCP port number, maybe?

**Comment by crawshaw** [comment]
General concept sounds good. I don't have time to review this morning.

File: ipn/ipnlocal/local.go [hunk 3]
```diff
@@ -1788,7 +1788,9 @@ func (b *LocalBackend) setPrefsLockedOnEntry(caller string, newp *ipn.Prefs) {
 	b.send(ipn.Notify{Prefs: newp})
 }
 
-func (b *LocalBackend) getPeerAPIPortForTSMPPing(ip netaddr.IP) (port uint16, ok bool) {
+// GetPeerAPIPort returns the port number for the peerapi server
+// running on the provided IP.
+func (b *LocalBackend) GetPeerAPIPort(ip netaddr.IP) (port uint16, ok bool) {
```
Comment (DentonGentry) [inline ipn/ipnlocal/local.go:1793]: So we bind to port 0 and let the kernel pick a port, but then pick off incoming packets to the port in netstack?

The main thing I'm trying to think through is if there could be a case where we close the kernel socket, potentially allowing some other process to bind to that same ephemeral port, but we neglect to cleanup the netstack listener and end up intercepting traffic which doesn't belong to us.

**Comment by maisem** [comment]
for posterity, why no iOS? we shouldn't be hitting memory limits anymore, so is it for battery usage?

File: ipn/ipnlocal/local.go (hunk 3 above)
Comment (bradfitz) [reply ipn/ipnlocal/local.go:1793]: that's why on every incoming SYN to netstack, we double-check that the PeerAPIPort is accurate.

**Comment by bradfitz** [comment]
> for posterity, why no iOS? we shouldn't be hitting memory limits anymore, so is it for battery usage?

Only iOS 15 gives us 50MB. Something like half of users are still on iOS 14.

We can enable it conditionally at runtime, but that doesn't help the binary size bloat, which still hurts us with our current version of Go due to the relocations. Josh's relocation work in the Go linker might make it possible for us to do that once we switch to Go 1.18.

**Review by JayWStapleton — COMMENTED** [review]
Tested and confirmed that it solved a recurring exit node DNS error we've been seeing.

File: ipn/ipnlocal/local.go (hunk 1 above)
Comment (bradfitz) [reply ipn/ipnlocal/local.go]: done

File: wgengine/netstack/netstack.go (hunk 2 above)
Comment (bradfitz) [reply wgengine/netstack/netstack.go]: reworded

File: ipn/ipnlocal/local.go (hunk 3 above)
Comment (bradfitz) [reply ipn/ipnlocal/local.go:1793]: > So we bind to port 0 and let the kernel pick a port, but then pick off incoming packets to the port in netstack?

Yes.

**Pull request merged**